File: packages/kt-client/src/main/kotlin/com/google/firebase/firestore/FieldPath.kt
package com.google.firebase.firestore

/**
 * A FieldPath refers to a field in a document. The path may consist of a single field name
 * (referring to a top-level field) or a list of field names (referring to a nested field).
 */
class FieldPath private constructor(val segments: List<String>) {

    init {
        require(segments.isNotEmpty()) { "Invalid field path. Path must not be empty." }
        for (segment in segments) {
            require(segment.isNotEmpty()) { "Invalid field name: must not be empty" }
        }
    }

    val canonicalString: String get() = segments.joinToString(".")

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is FieldPath) return false
        return segments == other.segments
    }

    override fun hashCode(): Int = segments.hashCode()

    override fun toString(): String = canonicalString

    companion object {
        private val DOCUMENT_ID_INSTANCE = FieldPath(listOf("__name__"))

        fun of(vararg fieldNames: String): FieldPath {
            require(fieldNames.isNotEmpty()) { "Invalid field path. Path must not be empty." }
            return FieldPath(fieldNames.toList())
        }

        fun fromDotSeparated(path: String): FieldPath {
            require(path.isNotEmpty()) { "Invalid field path. Path must not be empty." }
            return FieldPath(path.split("."))
        }

        fun documentId(): FieldPath = DOCUMENT_ID_INSTANCE

        fun extract(data: Map<String, Any?>?, path: String): Any? {
            if (data == null) return null
            if (data.containsKey(path)) return data[path]
            val segments = path.split(".")
            return extractBySegments(data, segments)
        }

        fun extract(data: Map<String, Any?>?, fieldPath: FieldPath): Any? {
            if (data == null) return null
            return extractBySegments(data, fieldPath.segments)
        }

        private fun extractBySegments(data: Map<String, Any?>?, segments: List<String>): Any? {
            var current: Any? = data
            for (segment in segments) {
                if (current !is Map<*, *>) return null
                current = current[segment]
            }
            return current
        }
    }
}

File: packages/swift-client/Sources/PyricFirestore/Public/FieldPath.swift
import Foundation

/// A FieldPath refers to a field in a document. The path may consist of a single field name
/// (referring to a top-level field in the document) or a list of field names (referring to a nested field).
public struct FieldPath: Equatable, Hashable, Sendable {
    public let segments: [String]

    public init(_ segments: [String]) {
        precondition(!segments.isEmpty, "FieldPath must not be empty.")
        for segment in segments {
            precondition(!segment.isEmpty, "FieldPath segments must not be empty.")
        }
        self.segments = segments
    }

    public init(field: String) {
        self.init([field])
    }

    public static func documentID() -> FieldPath {
        FieldPath(["__name__"])
    }

    public var stringValue: String {
        segments.joined(separator: ".")
    }
}

extension FieldPath: CustomStringConvertible {
    public var description: String {
        stringValue
    }
}
